src/lib.rs

```diff
@@ -102,48 +104,49 @@ impl MemoryListener {
     /// use memory_socket::MemoryListener;
     ///
     /// # fn main () -> ::std::io::Result<()> {
-    /// let listener = MemoryListener::bind(16)?;
+    /// let listener = MemoryListener::bind("192.51.100.2:1337")?;
     /// # Ok(())}
     /// ```
-    pub fn bind(port: u16) -> Result<Self> {
+    pub fn bind<A: ToSocketAddrs>(addresses: A) -> Result<Self> {
         let mut switchboard = (&*SWITCHBOARD).lock().unwrap();
 
-        // Get the port we should bind to.  If 0 was given, use a random port
-        let port = if let Some(port) = NonZeroU16::new(port) {
-            if switchboard.0.contains_key(&port) {
-                return Err(ErrorKind::AddrInUse.into());
-            }
-
-            port
-        } else {
-            loop {
-                let port = NonZeroU16::new(switchboard.1).unwrap_or_else(|| unreachable!());
+        let mut addresses = addresses.to_socket_addrs()?;
 
-                // The switchboard is full and all ports are in use
-                if switchboard.0.len() == (std::u16::MAX - 1) as usize {
-                    return Err(ErrorKind::AddrInUse.into());
-                }
+        let mut address = match addresses.next() {
+            Some(address) => address,
+            None => return Err(ErrorKind::AddrNotAvailable.into()),
+        };
 
-                // Instead of overflowing to 0, resume searching at port 1 since port 0 isn't a
-                // valid port to bind to.
-                if switchboard.1 == std::u16::MAX {
-                    switchboard.1 = 1;
-                } else {
-                    switchboard.1 += 1;
-                }
+        // It doesn't really make sense to listen on multiple interfaces in
+        // this environment, so we place a restriction on the parameter.
+        if addresses.next().is_some() {
+            return Err(ErrorKind::AddrNotAvailable.into());
+        }
+        // Similarly, it doesn't make a sense to listen on "all interfaces"
+        // in this environment, so return an error if they requested 0.0.0.0
+        // TODO: We could use get_if_addrs and use the host's real name?
+        if address.ip().is_unspecified() {
+            return Err(ErrorKind::AddrNotAvailable.into());
+        }
 
-                if !switchboard.0.contains_key(&port) {
-                    break port;
-                }
+        // If they didn't provide a port find one that isn't in use.
+        if address.port() == 0 {
+            let mut rng = thread_rng();
+            address.set_port(rng.gen());
+            while switchboard.0.contains_key(&address) {
+                address.set_port(rng.gen());
```

Comment:
While this probably won't happen in practice, but this could end up resulting in an infinite loop, unlike the current code which will bail when every port is used up. This also doesn't take into account that the fact that this could accidentally pass out port 0.